File: gyroTest/src/main/java/com/stulnikov/arduino/arduinoBt/gyrotest/gyrotest/MainActivity.java
package com.stulnikov.arduino.arduinoBt.gyrotest.gyrotest;

import android.bluetooth.BluetoothAdapter;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.SeekBar;
import android.widget.TextView;

import com.stulnikov.arduino.arduinoBt.gyrotest.gyrotest.dagger.DaggerPresenterComponent;

import java.util.Locale;

import javax.inject.Inject;

public class MainActivity extends AppCompatActivity implements MainView {

    public static final String TAG = "GyroTest";
    public static final int REQUEST_ENABLE_BT = 1;

    @Inject
    public MainPresenter mMainPresenter;

    private TextView mXValueText;
    private TextView mYValueText;
    private TextView mZValueText;
    private TextView mAngleValueText;
    private TextView mArduinoDataTextView;
    private View mContentLayout;
    private ProgressBar mProgressBar;

    private Button mRetryButton;

    private Button mRunButton;
    private SeekBar mRunProgress;
    private Button mRunBackButton;
    private Button mRobotButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initViews();
        if (mMainPresenter == null) {
            DaggerPresenterComponent
                    .builder()
                    .applicationComponent(((GyroTestApp) getApplication()).getApplicationComponent())
                    .build().inject(this);
        }
        mMainPresenter.bindView(this);
    }

    @Override
    protected void onStart() {
        super.onStart();
        mContentLayout.setVisibility(View.GONE);
        mRetryButton.setVisibility(View.GONE);
        mProgressBar.setVisibility(View.VISIBLE);
        mMainPresenter.start();
    }

    @Override
    protected void onStop() {
        super.onStop();
        mMainPresenter.stop();
    }

    @Override
    protected void onDestroy() {
        mMainPresenter.unbindView();
        super.onDestroy();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        return id == R.id.action_settings || super.onOptionsItemSelected(item);
    }

    @Override
    public void onDeviceConnected() {
        Log.d(TAG, "Device Connected");
        mArduinoDataTextView.setText(R.string.device_connected);
        mProgressBar.setVisibility(View.GONE);
        mRetryButton.setVisibility(View.GONE);
        mContentLayout.setVisibility(View.VISIBLE);
        mRobotButton.setVisibility(View.VISIBLE);
    }

    @Override
    public void onDeviceDisconnected() {
        Log.d(TAG, "Device Disconnected");
        mArduinoDataTextView.setText(R.string.device_disconnected);
        mProgressBar.setVisibility(View.GONE);
        mContentLayout.setVisibility(View.GONE);
        mRobotButton.setVisibility(View.GONE);
        mRetryButton.setVisibility(View.VISIBLE);
    }

    @Override
    public void showReceivedData(String data) {
        Log.d(TAG, "Data Received: " + data);
        if (!TextUtils.isEmpty(data)) {
            mArduinoDataTextView.setText(getString(R.string.data_received, data));
        }
    }

    @Override
    public void showEnableBluetooth() {
        Intent enableBtIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE);
        startActivityForResult(enableBtIntent, REQUEST_ENABLE_BT);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_ENABLE_BT) {
            if (resultCode == RESULT_OK) {
                mMainPresenter.start();
            } else {
                onDeviceDisconnected();
            }
        }
    }

    @Override
    public void toggleRobotMode(boolean robotMode) {
        mRunButton.setEnabled(!robotMode);
        mRunProgress.setEnabled(!robotMode);
        mRunBackButton.setEnabled(!robotMode);
    }

    @Override
    public void showAngle(float angle) {
        mAngleValueText.setText(getString(R.string.angle, angle));
    }

    @Override
    public void showX(float x) {
        mXValueText.setText(String.format(Locale.getDefault(), "%1.2f", x));
    }

    @Override
    public void showY(float y) {
        mYValueText.setText(String.format(Locale.getDefault(), "%1.2f", y));
    }

    @Override
    public void showZ(float z) {
        mZValueText.setText(String.format(Locale.getDefault(), "%1.2f", z));
    }

    private void initViews() {
        mContentLayout = findViewById(R.id.content_layout);
        mXValueText = (TextView) findViewById(R.id.value_x);
        mYValueText = (TextView) findViewById(R.id.value_y);
        mZValueText = (TextView) findViewById(R.id.value_z);
        mAngleValueText = (TextView) findViewById(R.id.angle_value);
        mArduinoDataTextView = (TextView) findViewById(R.id.arduino_data);

        mRunButton = (Button) findViewById(R.id.run_button);
        mRunButton.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        mMainPresenter.setDrive(true);
                        break;
                    case MotionEvent.ACTION_UP:
                        mMainPresenter.setDrive(false);
                        break;
                }
                return false;
            }
        });

        mRunProgress = findViewById(R.id.run_progress);
        mRunProgress.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                Log.d(TAG, "Action" + event.getAction());
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        // Do nothing
                        break;
                    case MotionEvent.ACTION_MOVE:
                        mMainPresenter.setDrivePercent(mRunProgress.getProgress());
                        break;
                    case MotionEvent.ACTION_UP:
                        mMainPresenter.setDrive(false);
                        break;
                }
                return false;
            }
        });

        mRunBackButton = (Button) findViewById(R.id.run_back_button);
        mRunBackButton.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        mMainPresenter.setDriveBack();
                        break;
                    case MotionEvent.ACTION_UP:
                        mMainPresenter.setDrive(false);
                        break;
                }
                return false;
            }
        });

        mRobotButton = (Button) findViewById(R.id.robot_button);
        mRobotButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mMainPresenter.toggleRobotMode();
            }
        });

        mProgressBar = (ProgressBar) findViewById(R.id.progress_bar);
        mRetryButton = (Button) findViewById(R.id.retry);
        mRetryButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Log.d(TAG, "Retry");
                mMainPresenter.start();
                mRetryButton.setVisibility(View.GONE);
                mProgressBar.setVisibility(View.VISIBLE);
            }
        });
    }
}

File: gyroTest/src/main/kotlin/com/stulnikov/arduino/arduinoBt/gyrotest/gyrotest/MainActivity.kt
package com.stulnikov.arduino.arduinoBt.gyrotest.gyrotest

import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.Menu
import android.view.MenuItem
import android.view.MotionEvent
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.stulnikov.arduino.arduinoBt.gyrotest.gyrotest.dagger.DaggerPresenterComponent
import kotlinx.android.synthetic.main.activity_main.*
import timber.log.Timber
import java.util.*
import javax.inject.Inject

class MainActivity : AppCompatActivity(), MainView {

    @Inject
    lateinit var mainPresenter: MainPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initViews()
        DaggerPresenterComponent
                .builder()
                .applicationComponent((application as GyroTestApp).applicationComponent)
                .build()
                .inject(this)
        mainPresenter.bindView(this)
    }

    override fun onStart() {
        super.onStart()
        content_layout.visibility = View.GONE
        retry.visibility = View.GONE
        progress_bar.visibility = View.VISIBLE
        mainPresenter.start()
    }

    override fun onStop() {
        super.onStop()
        mainPresenter.stop()
    }

    override fun onDestroy() {
        mainPresenter.unbindView()
        super.onDestroy()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        return id == R.id.action_settings || super.onOptionsItemSelected(item)
    }

    override fun onDeviceConnected() {
        Timber.d("Device Connected")
        arduino_data.setText(R.string.device_connected)
        progress_bar.visibility = View.GONE
        retry.visibility = View.GONE
        content_layout.visibility = View.VISIBLE
        robot_button.visibility = View.VISIBLE
    }

    override fun onDeviceDisconnected() {
        Timber.d("Device Disconnected")
        arduino_data.setText(R.string.device_disconnected)
        progress_bar.visibility = View.GONE
        content_layout.visibility = View.GONE
        robot_button.visibility = View.GONE
        retry.visibility = View.VISIBLE
    }

    override fun showReceivedData(data: String) {
        if (!TextUtils.isEmpty(data)) {
            arduino_data.text = getString(R.string.data_received, data)
        }
    }

    override fun showEnableBluetooth() {
        val enableBtIntent = Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)
        startActivityForResult(enableBtIntent, REQUEST_ENABLE_BT)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_ENABLE_BT) {
            if (resultCode == RESULT_OK) {
                mainPresenter.start()
            } else {
                onDeviceDisconnected()
            }
        }
    }

    override fun toggleRobotMode(robotMode: Boolean) {
        run_button.isEnabled = !robotMode
        run_progress.isEnabled = !robotMode
        run_back_button.isEnabled = !robotMode
    }

    override fun showAngle(angle: Float) {
        angle_value.text = getString(R.string.angle, angle)
    }

    override fun showX(x: Float) {
        value_x.text = String.format(Locale.getDefault(), "%1.2f", x)
    }

    override fun showY(y: Float) {
        value_y.text = String.format(Locale.getDefault(), "%1.2f", y)
    }

    override fun showZ(z: Float) {
        value_z.text = String.format(Locale.getDefault(), "%1.2f", z)
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun initViews() {
        run_button.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> mainPresenter.setDrive(true)
                MotionEvent.ACTION_UP -> mainPresenter.setDrive(false)
            }
            false
        }
        run_progress.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                }
                MotionEvent.ACTION_MOVE -> mainPresenter.setDrivePercent(run_progress.progress)
                MotionEvent.ACTION_UP -> mainPresenter.setDrive(false)
            }
            false
        }
        run_back_button.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> mainPresenter.setDriveBack()
                MotionEvent.ACTION_UP -> mainPresenter.setDrive(false)
            }
            false
        }
        robot_button.setOnClickListener { mainPresenter.toggleRobotMode() }
        retry.setOnClickListener {
            Timber.d("Retry")
            mainPresenter.start()
            retry.visibility = View.GONE
            progress_bar.visibility = View.VISIBLE
        }
    }

    override fun showFrontDistance(level: Int) {
        front_distance.setImageLevel(level)
    }

    override fun showRearDistance(level: Int) {
        rear_distance.setImageLevel(level)
    }

    companion object {
        const val REQUEST_ENABLE_BT = 1
    }
}